src/main/java/conversion/BaseServlet.java

```diff
@@ -100,22 +104,41 @@ protected void doPost(final HttpServletRequest request, final HttpServletRespons
 
             individual.isAlive = true;
 
-            final Part filePart = request.getPart("file");
-            if (filePart == null) {
+            final byte[] fileBytes;
+            // Attempt to get the filename from parameters (null if not in params).
+            String fileName = request.getParameter("filename");
+            Part filePart = null;
+            
+            // Fail silently now and check for null later as the file might be passed via a url.
+            try {
+                filePart = request.getPart("file");
+            } catch (ServletException e) { }
+
+            final String conversionUrl = request.getParameter("conversionUrl");
+            
+            // Prioritise file in request over file passed via url.
+            if (filePart != null) {
+                fileBytes = getFileFromRequestPart(filePart);
+                if (fileName == null) {
+                    fileName = getFileNameFromRequestPart(filePart);
+                }
+            } else if (conversionUrl != null) {
+                fileBytes = getFileFromUrl(conversionUrl, NUM_DOWNLOAD_RETRIES);
+                if (fileName == null) {
+                    fileName = getFileNameFromUrl(conversionUrl);
+                }
+            } else {
                 imap.remove(uuidStr);
                 doError(response, "Missing file", 400);
                 return;
             }
 
-            final byte[] fileBytes = new byte[(int) filePart.getSize()];
-            final InputStream fileContent = filePart.getInputStream();
-            fileContent.read(fileBytes);
-            fileContent.close();
-
-            String fileName = getFileName(filePart);
-            if (fileName == null) {
+            if (fileBytes == null) {
+                imap.remove(uuidStr);
+                doError(response, "Cannot get file data", 500);
+            } else if (fileName == null) {
                 imap.remove(uuidStr);
-                doError(response, "Missing file name", 500); // Would this ever occur?
+                doError(response, "Missing file", 500); // Would this ever occur?
```

Comment:
It could do, seeing as both getFileNameFrom...() methods can return null.

On a side note, please can we leave questions like this as comment on the pull request / commit rather than as a comment in the code